nexus/src/app/background/blueprint_execution.rs

```diff
@@ -0,0 +1,407 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for realizing a plan blueprint
+
+use super::common::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::stream;
+use futures::FutureExt;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::lookup::LookupPath;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::{Blueprint, OmicronZonesConfig};
+use serde_json::json;
+use sled_agent_client::Client as SledAgentClient;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Background task that takes a [`Blueprint`] and realizes the change to
+/// the state of the system based on the `Blueprint`.
+pub struct BlueprintExecutor {
+    datastore: Arc<DataStore>,
+    rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+}
+
+impl BlueprintExecutor {
+    // Temporary until we wire up the background task
+    #[allow(unused)]
+    pub fn new(
+        datastore: Arc<DataStore>,
+        rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+    ) -> BlueprintExecutor {
+        BlueprintExecutor { datastore, rx_blueprint }
+    }
+
+    // This is a modified copy of the functionality from `nexus/src/app/sled.rs`.
+    // There's no good way to access this functionality right now since it is a
+    // method on the `Nexus` type. We want to have a more constrained type we can
+    // pass into background tasks for this type of functionality, but for now we
+    // just copy the functionality.
+    async fn sled_client(
+        &self,
+        opctx: &OpContext,
+        sled_id: &Uuid,
+    ) -> Result<SledAgentClient, anyhow::Error> {
+        let (.., sled) = LookupPath::new(opctx, &self.datastore)
+            .sled_id(*sled_id)
+            .fetch()
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to create sled_agent::Client for sled_id: {}",
+                    sled_id
+                )
+            })?;
+        let dur = std::time::Duration::from_secs(60);
+        let client = reqwest::ClientBuilder::new()
+            .connect_timeout(dur)
+            .timeout(dur)
+            .build()
+            .unwrap();
+        Ok(SledAgentClient::new_with_client(
+            &format!("http://{}", sled.address()),
+            client,
+            opctx.log.clone(),
+        ))
+    }
+
+    async fn realize_blueprint(
+        &self,
+        opctx: &OpContext,
+        blueprint: &std::sync::Arc<Blueprint>,
```

Comment:
nit: I think you could accept `&Blueprint` here.

Reply:
Good catch!